spacy_transformers/layers/hf_shim.py

```diff
@@ -0,0 +1,95 @@
+from io import BytesIO
+from pathlib import Path
+from typing import Any
+import srsly
+from functools import partial
+import torch
+from spacy.vectors import get_current_ops
+
+from ..util import make_tempdir
+
+from thinc.api import PyTorchShim
+
+from transformers import AutoModel, AutoConfig, AutoTokenizer
+
+
+class HFModel:
+    def __init__(
+        self,
+        transformer,
+        tokenizer,
+        tokenizer_config: dict = {},
+        transformer_config: dict = {},
+    ):
+        self.transformer = transformer
+        self.tokenizer = tokenizer
+        self.tokenizer_config = tokenizer_config
+        self.transformer_config = transformer_config
+
+
+class HFShim(PyTorchShim):
+    """Interface between a HF Pytorch model and a Thinc Model."""
+
+    def __init__(self, model: HFModel, config=None, optimizer: Any = None):
+        self._hfmodel = model
+        super().__init__(model.transformer, config, optimizer)
+
+    def to_bytes(self):
+        config = {}
+        tok_dict = {}
+        weights_bytes = {}
+        hf_model = self._hfmodel
+        if hf_model.transformer is not None:
+            tok_dict = {}
+            config = hf_model.transformer.config.to_dict()
+            tokenizer = hf_model.tokenizer
+            with make_tempdir() as temp_dir:
+                tokenizer.save_pretrained(temp_dir)
+                for x in temp_dir.glob("**/*"):
+                    if x.is_file():
+                        tok_dict[x.name] = x.read_bytes()
+            filelike = BytesIO()
+            torch.save(self._model.state_dict(), filelike)
+            filelike.seek(0)
+            weights_bytes = filelike.getvalue()
+        msg = {
+            "config": config,
+            "state": weights_bytes,
+            "tokenizer": tok_dict,
+            "tokenizer_config": hf_model.tokenizer_config,
+            "transformer_config": hf_model.transformer_config,
+        }
+        return srsly.msgpack_dumps(msg)
+
+    def from_bytes(self, bytes_data):
+        msg = srsly.msgpack_loads(bytes_data)
+        config_dict = msg["config"]
+        tok_dict = msg["tokenizer"]
+        tok_config = msg["tokenizer_config"]
+        trf_config = msg["transformer_config"]
+        if config_dict:
+            with make_tempdir() as temp_dir:
+                config_file = temp_dir / "config.json"
+                srsly.write_json(config_file, config_dict)
+                config = AutoConfig.from_pretrained(config_file)
+                for x, x_bytes in tok_dict.items():
+                    Path(temp_dir / x).write_bytes(x_bytes)
```

Comment:
I like to do these inline opens, but did we have a problem recently where it left us with too many file handlers open in the CI sometimes? If it matters we would have to rewrite this to do the open in a context manager.

Reply:
Good point. I've been looking at pathlib's docs, and these methods should inherently open & close the file handlers themselves, eg https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_bytes (also https://stackoverflow.com/a/61444215/7961860 for instance)

Shall we risk it and keep the code clean, or do the context manager to be on the safe side?

Reply:
Actually, from the code:

```
    def read_bytes(self):
        """
        Open the file in bytes mode, read it, and close the file.
        """
        with self.open(mode='rb') as f:
            return f.read()

    def write_bytes(self, data):
        """
        Open the file in bytes mode, write to it, and close the file.
        """
        # type-check for the buffer interface before truncating the file
        view = memoryview(data)
        with self.open(mode='wb') as f:
            return f.write(view)
```
So I think we should be good.